azurelinuxagent/common/logcollector.py

```diff
@@ -0,0 +1,360 @@
+        # queue. The queue will determine the order in which the files are collected, highest priority files first.
+        priority_file_queue = []
+        for file in file_list:
+            priority = self._get_file_priority(file)
+            heappush(priority_file_queue, (priority, file))
+
+        return priority_file_queue
+
+    def _get_final_list_for_archive(self, priority_file_queue):
+        # Given a priority queue of files to collect, add one by one while the archive size is under the size limit.
+        # If a single file is over the file size limit, truncate it before adding it to the archive.
+        self._log_to_results_file("\n### Preparing list of files to add to archive ###")
+        total_uncompressed_size = 0
+        final_files_to_collect = []
+
+        while priority_file_queue:
+            file_path = heappop(priority_file_queue)[1]  # (priority, file_path)
+            file_size = min(os.path.getsize(file_path), FILE_SIZE_LIMIT)
+
+            if total_uncompressed_size + file_size > ARCHIVE_SIZE_LIMIT:
+                self._log_to_results_file("Archive too big, done with adding files.")
+                break
+
+            if os.path.getsize(file_path) <= FILE_SIZE_LIMIT:
+                final_files_to_collect.append(file_path)
+                self._log_to_results_file("Adding file {0}, size {1}b".format(file_path, file_size))
+            else:
+                truncated_file_path = self._truncate_large_file(file_path)
+                if truncated_file_path:
+                    self._log_to_results_file("Adding truncated file {0}, size {1}b".format(truncated_file_path, file_size))
+                    final_files_to_collect.append(truncated_file_path)
+
+            total_uncompressed_size += file_size
+
+        return final_files_to_collect
+
+    def _create_list_of_files_to_collect(self):
+        # The final list of files to be collected by zip is created in three steps:
+        # 1) Parse given manifest file, expanding wildcards and keeping a list of files that exist on disk
+        # 2) Assign those files a priority depending on whether they are in the must collect file list.
+        # 3) In priority order, add files to the final list to be collected, until the size of the archive is under
+        #    the size limit.
+        parsed_file_paths = self._parse_manifest_file()
+        prioritized_file_paths = self._get_priority_files_list(parsed_file_paths)
+        files_to_collect = self._get_final_list_for_archive(prioritized_file_paths)
+        return files_to_collect
+
+    def collect_logs(self):
+        """
+        Public method that collects necessary log files in a tarball that is updated each time this method is invoked.
+        The tarball is then compressed into a zip.
+        :return: Returns True if the log collection succeeded
+        """
+        try:
+            # Clear previous run's output and create base directories if they don't exist already
```

Comment:
you should clean up as soon as you're done creating the zip instead of cleaning on the next run

Reply:
The idea was to leave the previous run's output for debugging purposes. For example, in a situation where the tool is invoked and the archive couldn't be created, not even the results file will be added to the archive. In that case, there is value in leaving that file on disk instead of deleting it.